contracts/staking/test/unit_tests/staking_proxy_test.ts

```diff
@@ -0,0 +1,281 @@
+        [owner, authorizedAddress, ...notAuthorizedAddresses] = accounts;
+
+        // Deploy contracts
+        testContract = await TestProxyDestinationContract.deployFrom0xArtifactAsync(
+            artifacts.TestProxyDestination,
+            env.provider,
+            env.txDefaults,
+            artifacts,
+        );
+        testContract2 = await TestProxyDestinationContract.deployFrom0xArtifactAsync(
+            artifacts.TestProxyDestination,
+            env.provider,
+            env.txDefaults,
+            artifacts,
+        );
+        testProxyContract = await TestStakingProxyUnitContract.deployFrom0xArtifactAsync(
+            artifacts.TestStakingProxyUnit,
+            env.provider,
+            env.txDefaults,
+            artifacts,
+            testContract.address,
+        );
+        const logDecoderDependencies = _.mapValues(artifacts, v => v.compilerOutput.abi);
+        testContractViaProxy = new TestProxyDestinationContract(
+            testProxyContract.address,
+            env.provider,
+            env.txDefaults,
+            logDecoderDependencies,
+        );
+
+        // Add authorized address to Staking Proxy
+        await testProxyContract.addAuthorizedAddress.sendTransactionAsync(authorizedAddress, { from: owner });
+    });
+
+    describe('Fallback function', () => {
+        it('should pass back the return value of the destination contract', async () => {
+            const returnValue = await testContractViaProxy.echo.callAsync(testString);
+            expect(returnValue).to.equal(testString);
+        });
+
+        it('should revert with correct value when destination reverts', async () => {
+            return expect(testContractViaProxy.die.callAsync()).to.revertWith(testRevertString);
+        });
+
+        it('should revert if no staking contract is attached', async () => {
+            await testProxyContract.detachStakingContract.awaitTransactionSuccessAsync({ from: authorizedAddress });
+            const expectedError = new StakingRevertErrors.ProxyDestinationCannotBeNilError();
+            const tx = testContractViaProxy.echo.callAsync(testString);
+            return expect(tx).to.revertWith(expectedError);
+        });
+    });
+
+    describe('attachStakingContract', () => {
+        it('should successfully attaching a new staking contract', async () => {
+            // Cache existing staking contract and attach a new one
+            const initStakingContractAddress = await testProxyContract.stakingContract.callAsync();
+            const txReceipt = await testProxyContract.attachStakingContract.awaitTransactionSuccessAsync(
+                testContract2.address,
+                { from: authorizedAddress },
+            );
+
+            // Validate `ContractAttachedToProxy` event
+            expect(txReceipt.logs.length).to.be.gte(1);
+            const contractAttachedEvent: StakingProxyStakingContractAttachedToProxyEventArgs = (txReceipt
```

Comment:
This should be `const contractAttachedEvent = (txReceipt.logs[0] as LogWithDecodedArgs<StakingProxyStakingContractAttachedToProxyEventArgs>).args;` in order to be consistent with all of the other tests.

Reply:
we should adopt the convention of using `verifyEvents` from `test-utils` whenever we want to check event fields

Reply:
@mzhu25 this is great!